### PR TITLE
Update testing for 1.2, moving to 3.5 by default.

### PR DESCRIPTION
**Pull request by mhvk**
This is an alternative to #4225, in which I tried to make a mininal set and move to python 3. For this purpose, I made an explicit python 3.5 default, and removed the non-dependency tests on 2.7 and 3.5 (since those tests are already done with dependencies anyway). 

Now see if it actually passes...

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -42,61 +42,50 @@ matrix:
...
         - os: linux
-          env: PYTHON_VERSION=2.7 MAIN_CMD='pep8 astropy --count' SETUP_CMD=''
+          env: MAIN_CMD='pep8 astropy --count' SETUP_CMD=''
 
     allow_failures:
-      - env: PYTHON_VERSION=2.7 NUMPY_VERSION=dev SETUP_CMD='test'
+        # test with numpy development; we want to know when there are problems,
+        # but not necessarily be unable to move forward ourselves, so allow this
+        # to fail.
+        - env: NUMPY_VERSION=dev SETUP_CMD='test'
```
Comment (bsipocz) [inline .travis.yml]: you need to list this above in the matrix too, and list the duplicate of the exact same env call under ``allow_failures``.

Comment (mhvk) [inline .travis.yml]: Ah, I wondered about that duplication -- should have realised it had a reason!

Comment (bsipocz) [inline .travis.yml]: np, I probably spent way more time than what's healthy playing with travis :)

**Comment by bsipocz** [comment]
@mhvk - Sorry for the nitpicking, but I think a numpy 1.9 build would be needed, too.

**Comment by mhvk** [comment]
Yes, should have added numpy 1.9. Darn, still so many tests!

File: .travis.yml [hunk 2]
```diff
@@ -20,18 +20,19 @@ addons:
 env:
     global:
         # Set defaults to avoid repeating in most cases
-        - NUMPY_VERSION=1.9
+        # For now default python=3.4, since beautiful soup does not support 3.5
+        - PYTHON_VERSION=3.4
+        - NUMPY_VERSION=1.10
         - MAIN_CMD='python setup.py'
         - CONDA_DEPENDENCIES='Cython jinja2'
         - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautiful-soup'
```
Comment (astrofrog) [inline .travis.yml]: @mhvk - Python 3.5 does support beautiful soup, but you need to use the new package name, ``beautifulsoup4``

File: .travis.yml [hunk 3]
```diff
@@ -20,18 +20,19 @@ addons:
...
         - PIP_DEPENDENCIES=''
 
     matrix:
-        - PYTHON_VERSION=2.6 SETUP_CMD='egg_info'
+        - SETUP_CMD='egg_info'
         - PYTHON_VERSION=2.7 SETUP_CMD='egg_info'
         - PYTHON_VERSION=3.3 SETUP_CMD='egg_info'
         - PYTHON_VERSION=3.4 SETUP_CMD='egg_info'
-        - PYTHON_VERSION=3.5 SETUP_CMD='egg_info'
```
Comment (astrofrog) [inline .travis.yml]: I would advocate for this matrix above always listing the PYTHON_VERSION for clarity

**Comment by astrofrog** [comment]
@mhvk - can you fix the commit that says 1.12 to say 1.2 instead?

File: .travis.yml [hunk 4]
```diff
@@ -20,17 +20,18 @@ addons:
...
-        - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautiful-soup'
+        - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautifulsoup4'
         - PIP_DEPENDENCIES=''
 
     matrix:
+        - PYTHON_VERSION=3.5 SETUP_CMD='egg_info'
         - PYTHON_VERSION=2.7 SETUP_CMD='egg_info'
         - PYTHON_VERSION=3.3 SETUP_CMD='egg_info'
         - PYTHON_VERSION=3.4 SETUP_CMD='egg_info'
-        - PYTHON_VERSION=3.5 SETUP_CMD='egg_info'
```
Comment (astrofrog) [inline .travis.yml]: Can you change the order back to the original one?

**Comment by mhvk** [comment]
OK, both done.

File: .travis.yml [hunk 5]
```diff
@@ -20,10 +20,11 @@ addons:
 env:
     global:
         # Set defaults to avoid repeating in most cases
-        - NUMPY_VERSION=1.9
+        - PYTHON_VERSION=3.5
+        - NUMPY_VERSION=1.10
```
Comment (astrofrog) [inline .travis.yml]: Should we have ``NUMPY_VERSION=stable`` here to always update to the latest stable version?

**Comment by mhvk** [comment]
The sphinx documentation build on python3 seems to have quite a few different issues, including some documentation or code docstrings assuming python2, so I moved that back to python2 for now. See new issue #4481.

**Comment by mhvk** [comment]
It seems the main tests finally do not fail any more. Does anyone understand why coverage when down?

**Comment by bsipocz** [comment]
If you rebase it should work again. 

We saw this before (I just can't find the issue), and my work hypothesis was that if coverage is using an older than master base for this PR but compares it to master's coverage. Thus if the coverage increased between the base of this and master it takes it as it "decreased" with this PR.

**Comment by mhvk** [comment]
@bsipocz - OK, thanks, that sounds plausible.  Rebased now.

**Comment by mhvk** [comment]
@bsipocz - maybe plausible, but sadly not true. Looking closer at the coverage report, essentially all python files seem to be missing, and the travis `test --coverage` log is full of "cannot open source file", so there must be some path set wrong or so (see https://api.travis-ci.org/jobs/102154880/log.txt?deansi=true). Any suggestions?

**Comment by bsipocz** [comment]
@mhvk - maybe ``cpp-coveralls`` has some issues with python3? I'm just guessing, never saw this problem before.

**Comment by bsipocz** [comment]
Hmm, looking at the logs it seems that it was a problem before, too. Maybe @embray or @eteq have a good idea what is going on? Why aren't the c files found? Also why are they tried to be tested at all in ``cextern``?
https://travis-ci.org/astropy/astropy/jobs/102180694#L2309

**Comment by mhvk** [comment]
With #4645 in, so that numpy 1.10 is tested, I now returned to moving our tests to python3 by default.  It seems to work, except that the coverage decreased, in a way I find very hard to parse.  Any suggestions? Is this a problem?

Note that this keeps the sphinx build on python2; moving that to python3 is the goal of @bsipocz's #4556.

**Comment by bsipocz** [comment]
It seems that something goes wrong with the c code coveralls with the py3 run. I don't think that should be a blocker for this PR though, but maybe open a separate issue to tackle this.

**Comment by bsipocz** [comment]
However something is fishy with these summary numbers:

This PR: ``============ 7495 passed, 85 skipped, 69 xfailed in 1194.37 seconds ============``

Master: ``=========== 10295 passed, 115 skipped, 74 xfailed in 517.94 seconds ============``

**Comment by bsipocz** [comment]
I've started some debug runs on my travis: https://travis-ci.org/bsipocz/astropy/builds/113411787

**Comment by mhvk** [comment]
@bsipocz - OK, thanks!  If it is too unclear, I can also move coverage back to 2.7 for now, and raise a separate issue about that (as is already happening for sphinx...). I guess the upside to all of this is that we're finding and fixing python3 issues!

**Comment by bsipocz** [comment]
Well, I would even argue that we could leave the coverage with the py3 run, and fix the problem separately. Now it seems coverage unrelated, some of those tests involving the c files are not picked up (or something similar) in the py3 run.

**Comment by mhvk** [comment]
@bsipocz - Since it is only the C coverage that is affected, either just moving coverage to 3.5 and investigating further after, or keeping it at 2.7 for now and moving later are fine by me. Maybe get further opinions... @astrofrog? @eteq?

**Comment by bsipocz** [comment]
@mhvk - OK, one mystery is solved. The difference in the number of tests has nothing to do with the c coverage, and due to ``# TEST_UNICODE_LITERALS``, in python2 those modules are run twice.

**Comment by eteq** [comment]
Oh, interesting.  So I guess this is our "real" test coverage then?  :( Perhaps we need an improve-coverage sprint in the near future...

At any rate, I think this is all set, right? @astrofrog, do you agree we want to do this? (i.e., switch to 3.5 as the default) If so, I say go for it!

**Comment by saimn** [comment]
Something seems wrong with this coverage report : `5842 of 8459 relevant lines covered` ?
On a report from another PR with Python 2 it says `42562 of 55530
relevant lines covered` which seems more realistic.
(`cloc --exclude-dir=tests,extern,compat astropy` reports 60873 lines of Python code).

**Comment by astrofrog** [comment]
By the way, I fixed issues with coverage in Python 3.x in https://github.com/astropy/astropy/pull/4822

**Comment by astrofrog** [comment]
However, as mentioned in https://github.com/astropy/astropy/issues/4826, coverage testing is way slower in Python 3, so I don't think we want to use that yet until this is figured out.

I'm removing the 1.2 milestone since this is not critical for 1.2

**Comment by mhvk** [comment]
Enough had changed in `.travis.yml` that I thought it was better to start afresh with trying to move our testing to python3; see #5155.
